### PR TITLE
Move union creation to a different class

### PR DESCRIPTION
**Pull request by riyafa**
Resolve https://github.com/ballerina-platform/ballerina-lang/issues/23427

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmUnionTypeConstantsGen.java [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+import static org.objectweb.asm.Opcodes.INVOKESTATIC;
+import static org.objectweb.asm.Opcodes.RETURN;
+import static org.objectweb.asm.Opcodes.V1_8;
+
+/**
+ * Generates Jvm class for the ballerina union types as constants for a given module.
+ *
+ * @since 2.0.0
+ */
+public class JvmUnionTypeConstantsGen {
```
Comment (warunalakshitha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmUnionTypeConstantsGen.java:52]: Shall we check with following sample.

```
enum MyZnum {

    Z1,
    Z2,
    Z3,
    Z4,
    Z5,
    Z6,
    Z7,
    Z8,
    Z9,
    Z10,
    Z11,
    Z12,
    Z13,
    Z14,
    Z15,
    Z16,
    Z17,
    Z18,
    Z19,
    Z20,
    Z21,
    Z22,
    Z23,
    Z24,
    Z25,
    Z26,
    Z27,
    Z28,
    Z29,
    Z30,
    Z31,
    Z32,
    Z33,
    Z34,
    Z35,
    Z36,
    Z37,
    Z38,
    Z39,
    Z40,
    Z41,
    Z42,
    Z43,
    Z44,
    Z45,
    Z46,
    Z47,
    Z48,
    Z49,
    Z50,
    Z51,
    Z52,
    Z53,
    Z54,
    Z55,
    Z56,
    Z57,
    Z58,
    Z59,
    Z60,
    Z61,
    Z62,
    Z63,
    Z64,
    Z65,
    Z66,
    Z67,
    Z68,
    Z69,
    Z70,
    Z71,
    Z72,
    Z73,
    Z74,
    Z75,
    Z76,
    Z77,
    Z78,
    Z79,
    Z80,
    Z81,
    Z82,
    Z83,
    Z84,
    Z85,
    Z86,
    Z87,
    Z88,
    Z89,
    Z90,
    Z91,
    Z92,
    Z93,
    Z94,
    Z95,
    Z96,
    Z97,
    Z98,
    Z99,
    Z100,
    Z101,
    Z102,
    Z103,
    Z104,
    Z105,
    Z106,
    Z107,
    Z108,
    Z109,
    Z110,
    Z111,
    Z112,
    Z113,
    Z114,
    Z115,
    Z116,
    Z117,
    Z118,
    Z119,
    Z120,
    Z121,
    Z122,
    Z123,
    Z124,
    Z125,
    Z126,
    Z127,
    Z128,
    Z129,
    Z130,
    Z131,
    Z132,
    Z133,
    Z134,
    Z135,
    Z136,
    Z137,
    Z138,
    Z139,
    Z140,
    Z141,
    Z142,
    Z143,
    Z144,
    Z145,
    Z146,
    Z147,
    Z148,
    Z149,
    Z150,
    Z151,
    Z152,
    Z153,
    Z154,
    Z155,
    Z156,
    Z157,
    Z158,
    Z159,
    Z160,
    Z161,
    Z162,
    Z163,
    Z164,
    Z165,
    Z166,
    Z167,
    Z168,
    Z169,
    Z170,
    Z171,
    Z172,
    Z173,
    Z174,
    Z175,
    Z176,
    Z177,
    Z178,
    Z179,
    Z180,
    Z181,
    Z182,
    Z183,
    Z184,
    Z185,
    Z186,
    Z187,
    Z188,
    Z189,
    Z190,
    Z191,
    Z192,
    Z193,
    Z194,
    Z195,
    Z196,
    Z197,
    Z198,
    Z199,
    Z200,
    Z201,
    Z202,
    Z203,
    Z204,
    Z205,
    Z206,
    Z207,
    Z208,
    Z209,
    Z210,
    Z211,
    Z212,
    Z213,
    Z214,
    Z215,
    Z216,
    Z217,
    Z218,
    Z219,
    Z220,
    Z221,
    Z222,
    Z223,
    Z224,
    Z225,
    Z226,
    Z227,
    Z228,
    Z229,
    Z230,
    Z231,
    Z232,
    Z233,
    Z234,
    Z235,
    Z236,
    Z237,
    Z238,
    Z239,
    Z240,
    Z241,
    Z242,
    Z243,
    Z244,
    Z245,
    Z246,
    Z247,
    Z248,
    Z249,
    Z250,
    Z251,
    Z252,
    Z253,
    Z254,
    Z255,
    Z256,
    Z257,
    Z258,
    Z259,
    Z260,
    Z261,
    Z262,
    Z263,
    Z264,
    Z265,
    Z266,
    Z267,
    Z268,
    Z269,
    Z270,
    Z271,
    Z272,
    Z273,
    Z274,
    Z275,
    Z276,
    Z277,
    Z278,
    Z279,
    Z280,
    Z281,
    Z282,
    Z283,
    Z284,
    Z285,
    Z286,
    Z287,
    Z288,
    Z289,
    Z290,
    Z291,
    Z292,
    Z293,
    Z294,
    Z295,
    Z296,
    Z297,
    Z298,
    Z299,
    Z300,
    Z301,
    Z302,
    Z303,
    Z304,
    Z305,
    Z306,
    Z307,
    Z308,
    Z309,
    Z310,
    Z311,
    Z312,
    Z313,
    Z314,
    Z315,
    Z316,
    Z317,
    Z318,
    Z319,
    Z320,
    Z321,
    Z322,
    Z323,
    Z324,
    Z325,
    Z326,
    Z327,
    Z328,
    Z329,
    Z330,
    Z331,
    Z332,
    Z333,
    Z334,
    Z335,
    Z336,
    Z337,
    Z338,
    Z339,
    Z340,
    Z341,
    Z342,
    Z343,
    Z344,
    Z345,
    Z346,
    Z347,
    Z348,
    Z349,
    Z350,
    Z351,
    Z352,
    Z353,
    Z354,
    Z355,
    Z356,
    Z357,
    Z358,
    Z359,
    Z360,
    Z361,
    Z362,
    Z363,
    Z364,
    Z365,
    Z366,
    Z367,
    Z368,
    Z369,
    Z370,
    Z371,
    Z372,
    Z373,
    Z374,
    Z375,
    Z376,
    Z377,
    Z378,
    Z379,
    Z380,
    Z381,
    Z382,
    Z383,
    Z384,
    Z385,
    Z386,
    Z387,
    Z388,
    Z389,
    Z390,
    Z391,
    Z392,
    Z393,
    Z394,
    Z395,
    Z396,
    Z397,
    Z398,
    Z399,
    Z400,
    Z401,
    Z402,
    Z403,
    Z404,
    Z405,
    Z406,
    Z407,
    Z408,
    Z409,
    Z410,
    Z411,
    Z412,
    Z413,
    Z414,
    Z415,
    Z416,
    Z417,
    Z418,
    Z419,
    Z420,
    Z421,
    Z422,
    Z423,
    Z424,
    Z425,
    Z426,
    Z427,
    Z428,
    Z429,
    Z430,
    Z431,
    Z432,
    Z433,
    Z434,
    Z435,
    Z436,
    Z437,
    Z438,
    Z439,
    Z440,
    Z441,
    Z442,
    Z443,
    Z444,
    Z445,
    Z446,
    Z447,
    Z448,
    Z449,
    Z450,
    Z451,
    Z452,
    Z453,
    Z454,
    Z455,
    Z456,
    Z457,
    Z458,
    Z459,
    Z460,
    Z461,
    Z462,
    Z463,
    Z464,
    Z465,
    Z466,
    Z467,
    Z468,
    Z469,
    Z470,
    Z471,
    Z472,
    Z473,
    Z474,
    Z475,
    Z476,
    Z477,
    Z478,
    Z479,
    Z480,
    Z481,
    Z482,
    Z483,
    Z484,
    Z485,
    Z486,
    Z487,
    Z488,
    Z489,
    Z490,
    Z491,
    Z492,
    Z493,
    Z494,
    Z495,
    Z496,
    Z497,
    Z498,
    Z499,
    Z500,
    Z501,
    Z502,
    Z503,
    Z504,
    Z505,
    Z506,
    Z507,
    Z508,
    Z509,
    Z510,
    Z511,
    Z512,
    Z513,
    Z514,
    Z515,
    Z516,
    Z517,
    Z518,
    Z519,
    Z520,
    Z521,
    Z522,
    Z523,
    Z524,
    Z525,
    Z526,
    Z527,
    Z528,
    Z529,
    Z530,
    Z531,
    Z532,
    Z533,
    Z534,
    Z535,
    Z536,
    Z537,
    Z538,
    Z539,
    Z540,
    Z541,
    Z542,
    Z543,
    Z544,
    Z545,
    Z546,
    Z547,
    Z548,
    Z549,
    Z550,
    Z551,
    Z552,
    Z553,
    Z554,
    Z555,
    Z556,
    Z557,
    Z558,
    Z559,
    Z560,
    Z561,
    Z562,
    Z563,
    Z564,
    Z565,
    Z566,
    Z567,
    Z568,
    Z569,
    Z570,
    Z571,
    Z572,
    Z573,
    Z574,
    Z575,
    Z576,
    Z577,
    Z578,
    Z579,
    Z580,
    Z581,
    Z582,
    Z583,
    Z584,
    Z585,
    Z586,
    Z587,
    Z588,
    Z589,
    Z590,
    Z591,
    Z592,
    Z593,
    Z594,
    Z595,
    Z596,
    Z597,
    Z598,
    Z599,
    Z600,
    Z601,
    Z602,
    Z603,
    Z604,
    Z605,
    Z606,
    Z607,
    Z608,
    Z609,
    Z610,
    Z611,
    Z612,
    Z613,
    Z614,
    Z615,
    Z616,
    Z617,
    Z618,
    Z619,
    Z620,
    Z621,
    Z622,
    Z623,
    Z624,
    Z625,
    Z626,
    Z627,
    Z628,
    Z629,
    Z630,
    Z631,
    Z632,
    Z633,
    Z634,
    Z635,
    Z636,
    Z637,
    Z638,
    Z639,
    Z640,
    Z641,
    Z642,
    Z643,
    Z644,
    Z645,
    Z646,
    Z647,
    Z648,
    Z649,
    Z650,
    Z651,
    Z652,
    Z653,
    Z654,
    Z655,
    Z656,
    Z657,
    Z658,
    Z659,
    Z660,
    Z661,
    Z662,
    Z663,
    Z664,
    Z665,
    Z666,
    Z667,
    Z668,
    Z669,
    Z670,
    Z671,
    Z672,
    Z673,
    Z674,
    Z675,
    Z676,
    Z677,
    Z678,
    Z679,
    Z680,
    Z681,
    Z682,
    Z683,
    Z684,
    Z685,
    Z686,
    Z687,
    Z688,
    Z689,
    Z690,
    Z691,
    Z692,
    Z693,
    Z694,
    Z695,
    Z696,
    Z697,
    Z698,
    Z699,
    Z700,
    Z701,
    Z702,
    Z703,
    Z704,
    Z705,
    Z706,
    Z707,
    Z708,
    Z709,
    Z710,
    Z711,
    Z712,
    Z713,
    Z714,
    Z715,
    Z716,
    Z717,
    Z718,
    Z719,
    Z720,
    Z721,
    Z722,
    Z723,
    Z724,
    Z725,
    Z726,
    Z727,
    Z728,
    Z729,
    Z730,
    Z731,
    Z732,
    Z733,
    Z734,
    Z735,
    Z736,
    Z737,
    Z738,
    Z739,
    Z740,
    Z741,
    Z742,
    Z743,
    Z744,
    Z745,
    Z746,
    Z747,
    Z748,
    Z749,
    Z750,
    Z751,
    Z752,
    Z753,
    Z754,
    Z755,
    Z756,
    Z757,
    Z758,
    Z759,
    Z760,
    Z761,
    Z762,
    Z763,
    Z764,
    Z765,
    Z766,
    Z767,
    Z768,
    Z769,
    Z770,
    Z771,
    Z772,
    Z773,
    Z774,
    Z775,
    Z776,
    Z777,
    Z778,
    Z779,
    Z780,
    Z781,
    Z782,
    Z783,
    Z784,
    Z785,
    Z786,
    Z787,
    Z788,
    Z789,
    Z790,
    Z791,
    Z792,
    Z793,
    Z794,
    Z795,
    Z796,
    Z797,
    Z798,
    Z799,
    Z800,
    Z801,
    Z802,
    Z803,
    Z804,
    Z805,
    Z806,
    Z807,
    Z808,
    Z809,
    Z810,
    Z811,
    Z812,
    Z813,
    Z814,
    Z815,
    Z816,
    Z817,
    Z818,
    Z819,
    Z820,
    Z821,
    Z822,
    Z823,
    Z824,
    Z825,
    Z826,
    Z827,
    Z828,
    Z829,
    Z830,
    Z831,
    Z832,
    Z833,
    Z834,
    Z835,
    Z836,
    Z837,
    Z838,
    Z839,
    Z840,
    Z841,
    Z842,
    Z843,
    Z844,
    Z845,
    Z846,
    Z847,
    Z848,
    Z849,
    Z850,
    Z851,
    Z852,
    Z853,
    Z854,
    Z855,
    Z856,
    Z857,
    Z858,
    Z859,
    Z860,
    Z861,
    Z862,
    Z863,
    Z864,
    Z865,
    Z866,
    Z867,
    Z868,
    Z869,
    Z870,
    Z871,
    Z872,
    Z873,
    Z874,
    Z875,
    Z876,
    Z877,
    Z878,
    Z879,
    Z880,
    Z881,
    Z882,
    Z883,
    Z884,
    Z885,
    Z886,
    Z887,
    Z888,
    Z889,
    Z890,
    Z891,
    Z892,
    Z893,
    Z894,
    Z895,
    Z896,
    Z897,
    Z898,
    Z899,
    Z900,
    Z901,
    Z902,
    Z903,
    Z904,
    Z905,
    Z906,
    Z907,
    Z908,
    Z909,
    Z910,
    Z911,
    Z912,
    Z913,
    Z914,
    Z915,
    Z916,
    Z917,
    Z918,
    Z919,
    Z920,
    Z921,
    Z922,
    Z923,
    Z924,
    Z925,
    Z926,
    Z927,
    Z928,
    Z929,
    Z930,
    Z931,
    Z932,
    Z933,
    Z934,
    Z935,
    Z936,
    Z937,
    Z938,
    Z939,
    Z940,
    Z941,
    Z942,
    Z943,
    Z944,
    Z945,
    Z946,
    Z947,
    Z948,
    Z949,
    Z950,
    Z951,
    Z952,
    Z953,
    Z954,
    Z955,
    Z956,
    Z957,
    Z958,
    Z959,
    Z960,
    Z961,
    Z962,
    Z963,
    Z964,
    Z965,
    Z966,
    Z967,
    Z968,
    Z969,
    Z970,
    Z971,
    Z972,
    Z973,
    Z974,
    Z975,
    Z976,
    Z977,
    Z978,
    Z979,
    Z980,
    Z981,
    Z982,
    Z983,
    Z984,
    Z985,
    Z986,
    Z987,
    Z988,
    Z989,
    Z990,
    Z991,
    Z992,
    Z993,
    Z994,
    Z995,
    Z996,
    Z997,
    Z998,
    Z999,
    Z1000
}

```

**Review by warunalakshitha — COMMENTED** [review]
LGTM

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmUnionTypeConstantsGen.java [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    private final Deque<BUnionType> stack;
+
+    private static final int MAX_MEMBERS_PER_METHOD = 100;
+
+    public JvmUnionTypeConstantsGen(PackageID packageID) {
+        unionTypeVarMap = new ConcurrentSkipListMap<>(this::checkUnionEqualityInInts);
+        unionVarConstantsClass = JvmCodeGenUtil.getModuleLevelClassName(
+                packageID, JvmConstants.BUNION_TYPE_CONSTANT_CLASS_NAME);
+        stack = new LinkedList<>();
+        generateClassInit();
```
Comment (HindujaB) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmUnionTypeConstantsGen.java:74]: The method names `visitMethod` & `generateClassInit` seems to be generic. Shall we have something like `visitUnionInitMethod()` & `genUnionConstantsClassInit()`

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmConstants.java [hunk 3]
```diff
@@ -264,13 +265,15 @@
...
     public static final String MODULE_STRING_CONSTANT_CLASS_NAME = "$_string_constants";
+    public static final String BUNION_TYPE_CONSTANT_CLASS_NAME = "$_bunion_type_constants";
     public static final String MODULE_TYPES_CLASS_NAME = "types/$_types";
     public static final String MODULE_ANON_TYPES_CLASS_NAME = "types/$_anon_types";
     public static final String MODULE_RECORDS_CLASS_NAME = "types/$_records";
     public static final String MODULE_OBJECTS_CLASS_NAME = "types/$_objects";
     public static final String MODULE_ERRORS_CLASS_NAME = "types/$_errors";
     public static final String MODULE_ANNOTATIONS_CLASS_NAME = "annotations/$_annotations";
     public static final String B_STRING_INIT_METHOD_PREFIX = "$string_init";
+    public static final String B_UNION_TYPE_INIT_METHOD_PREFIX = "$bunion_type_init";
```
Comment (warunalakshitha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmConstants.java:276]:
```suggestion
    public static final String B_UNION_TYPE_INIT_METHOD_PREFIX = "$union_type_init";
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmUnionTypeConstantsGen.java [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+    private ClassWriter cw;
+    private MethodVisitor mv;
+    private int memberCountInMethod;
+    private int methodCount;
+    /**
+     * Stack keeps track of recursion in union types. The method creation is performed only if recursion is completed.
+     */
+    private final Deque<BUnionType> stack;
+
+    private static final int MAX_MEMBERS_PER_METHOD = 50;
```
Comment (warunalakshitha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmUnionTypeConstantsGen.java:67]: Can we make this 100?

Comment (warunalakshitha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmUnionTypeConstantsGen.java:67]: Our type loading will have 100 as well. that mean it each type load can have 6400 instructions that should be enough IMO.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmUnionTypeConstantsGen.java [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+        }
+        if (o1.getMemberTypes().size() != o2.getMemberTypes().size() || !o1.toString().equals(o2.toString())) {
+            return false;
+        }
+        for (BType type : o1.getMemberTypes()) {
+            if (!o2.getMemberTypes().contains(type)) {
+                return false;
+            }
+        }
+
```
Comment (warunalakshitha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmUnionTypeConstantsGen.java:101]: Do we need this empty line?

File: tests/jballerina-unit-test/src/test/resources/test-src/jvm/largePackage/main.bal [hunk 6]
```diff
@@ -37,4 +37,4 @@ public function main() {
     e:MyError1 myError = error("My Error");
     test:assertEquals(myError.message(), "My Error");
     test:assertEquals(u:v1.a, "hello");
```
Comment (warunalakshitha) [inline tests/jballerina-unit-test/src/test/resources/test-src/jvm/largePackage/main.bal:39]: Shall we use big enum in main.bal

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmUnionTypeConstantsGen.java (hunk 5 above)
Comment (riyafa) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmUnionTypeConstantsGen.java]: removed it

**Pull request closed without merging**